radix-engine/src/kernel/track.rs

```diff
@@ -703,3 +717,288 @@ impl<'s> FinalizingTrack<'s> {
+                        }
+                        BalanceChange::NonFungible { added, removed } => {
+                            let existing = balance_changes
+                                .entry(*root)
+                                .or_default()
+                                .entry(resource_address)
+                                .or_insert(BalanceChange::NonFungible {
+                                    added: BTreeSet::new(),
+                                    removed: BTreeSet::new(),
+                                });
+                            existing.added_non_fungibles().extend(added);
+                            existing.removed_non_fungibles().extend(removed);
+                        }
+                    }
+                }
+            } else {
+                // Scan loaded substates to find children
+                for (_module_id, module_substates) in modules {
+                    for (_, update) in module_substates {
+                        let substate_value = IndexedScryptoValue::from_typed(&update.0);
+                        for own in substate_value.owned_node_ids() {
+                            self.traverse_state_updates(
+                                balance_changes,
+                                accounted_vaults,
+                                root,
+                                own,
+                            );
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    fn is_vault(&self, node_id: &RENodeId) -> bool {
+        let type_info = self
+            .fetch_substate(&SubstateId(
+                *node_id,
+                NodeModuleId::TypeInfo,
+                SubstateOffset::TypeInfo(TypeInfoOffset::TypeInfo),
+            ))
+            .type_info()
+            .clone();
+
+        type_info.package_address == RESOURCE_MANAGER_PACKAGE
+            && type_info.blueprint_name == VAULT_BLUEPRINT
+    }
+
+    fn calculate_vault_balance_change(
+        &self,
+        node_id: &RENodeId,
+    ) -> Option<(ResourceAddress, BalanceChange)> {
+        let vault_info = self
+            .fetch_substate(&SubstateId(
+                *node_id,
+                NodeModuleId::SELF,
+                SubstateOffset::Vault(VaultOffset::Info),
+            ))
+            .vault_info()
+            .clone();
+
+        if vault_info.resource_type.is_fungible() {
+            // If there is an update to the liquid resource
+            if let Some((substate, old_version)) =
+                self.fetch_substate_from_state_updates(&SubstateId(
+                    *node_id,
+                    NodeModuleId::SELF,
+                    SubstateOffset::Vault(VaultOffset::LiquidFungible),
+                ))
+            {
+                let old_balance = if old_version.is_none() {
+                    Decimal::ZERO
+                } else {
+                    self.fetch_substate_from_store(&SubstateId(
```

Comment:
Ahh don't we store the old version in the substate updates? Maybe we should do some kind of CoW for substates which have a mutable lock taken.

(Not something to change as part of this PR!)

Reply:
We currently don't, which is why I need to reload the substate from `substate_store`.

A better approach is probabaly to keep a copy of the original value in the track. Will consider this during track refactoring.

Reply:
Agreed, that makes sense 👍